src/internal/common/network/retry_handler.go

```diff
@@ -0,0 +1,186 @@
+package network
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"syscall"
+
+	"github.com/alcionai/clues"
+)
+
+var _ io.ReadCloser = &resetRetryHandler{}
+
+const (
+	numMaxRetries  = 3
+	rangeHeaderKey = "Range"
+	// One-sided range like this is defined as starting at the given byte and
+	// extending to the end of the item.
+	rangeHeaderOneSidedValueTmpl = "bytes=%d-"
+)
+
+// Could make this per wrapper instance if we need additional flexibility
+// between callers.
+var retryErrs = []error{
+	syscall.ECONNRESET,
+}
+
+type Getter interface {
+	// SupportsRangeReq returns true if this Getter supports adding Range headers
+	// to the Get call. Otherwise returns false.
+	SupportsRangeReq() bool
+	// Get attempts to get another reader for the data this reader is returning.
+	// headers denotes any additional headers that should be added to the request,
+	// like a Range header.
+	//
+	// Don't allow passing a URL to Get so that we can hide the fact that some
+	// components may need to dynamically refresh the fetch URL (i.e. OneDrive)
+	// from this wrapper.
+	//
+	// Get should encapsulate all error handling and status code checking required
+	// for the component. This function is called both during NewResetRetryHandler
+	// and Read so it's possible to discover errors with the item prior to
+	// informing other components about it if desired.
+	Get(ctx context.Context, headers map[string]string) (io.ReadCloser, error)
+}
+
+func NewResetRetryHandler(
+	ctx context.Context,
+	getter Getter,
+) (*resetRetryHandler, error) {
+	rrh := &resetRetryHandler{
+		ctx:    ctx,
+		getter: getter,
+	}
+
+	// Retry logic encapsulated in reconnect so no need for it here.
+	_, err := rrh.reconnect(numMaxRetries)
+
+	return rrh, clues.Wrap(err, "initializing reader").OrNil()
+}
+
+//nolint:unused
+type resetRetryHandler struct {
+	ctx         context.Context
+	getter      Getter
+	innerReader io.ReadCloser
+	offset      int64
+}
+
+func isRetriable(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	for _, e := range retryErrs {
+		if errors.Is(err, e) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (rrh *resetRetryHandler) Read(p []byte) (int, error) {
+	if rrh.innerReader == nil {
+		return 0, clues.New("not initialized")
+	}
+
+	var (
+		// Use separate error variable just to make other assignments in the loop a
+		// bit cleaner.
+		finalErr   error
+		read       int
+		numRetries int
+	)
+
+	// Still need to check retry count in loop header so we don't go through one
+	// last time after failing to reconnect due to exhausting retries.
+	for numRetries < numMaxRetries {
+		n, err := rrh.innerReader.Read(p[read:])
+		rrh.offset = rrh.offset + int64(n)
+		read = read + n
+
+		// Catch short reads with no error and errors we don't know how to retry.
+		if !isRetriable(err) || numRetries >= numMaxRetries {
```

Comment:
nit: `numRetries >= numMaxRetries` will always be false here due to check on line 100.